File: frontend/src/types/market.ts
/**
 * TypeScript types for Phase 5 Market Intelligence API.
 */

export interface LocationSearchResult {
  display_name: string
  latitude:     number
  longitude:    number
  place_id?:    string
  country?:     string
  state?:       string
  district?:    string
}

export interface MarketAnalyzeRequest {
  business_id: string
  latitude:    number
  longitude:   number
  radius_km?:  number
}

export interface NearbyRequest {
  latitude:  number
  longitude: number
  radius_km?: number
}

export interface NearbyPlaceOut {
  osm_id:          string
  name:            string
  category:        string
  latitude:        number
  longitude:       number
  distance_meters: number
}

export interface CompetitorSummaryOut {
  direct_count:      number
  related_count:     number
  total_businesses:  number
  competition_level: 'Low' | 'Moderate' | 'High'
  density_per_sqkm:  number
}

export interface MarketOpportunityOut {
  competition_score:    number
  infrastructure_score: number
  accessibility_score:  number
  diversity_score:      number
  market_size_score:    number
  total:                number
}

export interface LocationSuitabilityOut {
  competition_score:    number
  infrastructure_score: number
  customer_proxy_score: number
  business_density:     number
  total:                number
}

export interface MarketInsightOut {
  icon:    string
  message: string
  level:   'positive' | 'warning' | 'neutral'
}

export interface MarketAnalysisOut {
  latitude:            number
  longitude:           number
  radius_km:           number
  location_name:       string | null
  business_name:       string
  business_id:         string
  competitor_summary:  CompetitorSummaryOut
  opportunity:         MarketOpportunityOut
  suitability:         LocationSuitabilityOut
  nearby_places:       NearbyPlaceOut[]
  direct_competitors:  NearbyPlaceOut[]
  insights:            MarketInsightOut[]
  recommendations:     string[]
  disclaimer:          string
}

export interface NearbyResultOut {
  latitude:     number
  longitude:    number
  radius_km:    number
  total_places: number
  places:       NearbyPlaceOut[]
  categories:   Record<string, number>
  from_cache:   boolean
  error:        string | null
  disclaimer:   string
}

File: backend/app/schemas/market.py
"""
Pydantic schemas for Phase 5 Market Intelligence API.
"""
from typing import Dict, List, Optional
from pydantic import BaseModel, Field, field_validator


# ─────────────────────────────────────────────────────────────────────────────
# Request schemas
# ─────────────────────────────────────────────────────────────────────────────

class MarketAnalyzeRequest(BaseModel):
    """POST /market/analyze"""
    business_id: str           = Field(..., description="Business ID from the catalogue")
    latitude:    float         = Field(..., ge=-90, le=90,   description="Location latitude")
    longitude:   float         = Field(..., ge=-180, le=180, description="Location longitude")
    radius_km:   float         = Field(5.0, ge=1, le=15,    description="Search radius in km (1–15)")

    @field_validator("radius_km")
    @classmethod
    def round_radius(cls, v: float) -> float:
        # Snap to nearest allowed value for cleaner UX
        allowed = [1, 2, 5, 10]
        return min(allowed, key=lambda x: abs(x - v))


class NearbyRequest(BaseModel):
    """POST /market/nearby"""
    latitude:  float = Field(..., ge=-90,  le=90)
    longitude: float = Field(..., ge=-180, le=180)
    radius_km: float = Field(5.0, ge=1, le=15)


# ─────────────────────────────────────────────────────────────────────────────
# Response schemas
# ─────────────────────────────────────────────────────────────────────────────

class LocationSearchResult(BaseModel):
    display_name: str
    latitude:     float
    longitude:    float
    place_id:     Optional[str] = None
    country:      Optional[str] = None
    state:        Optional[str] = None
    district:     Optional[str] = None


class NearbyPlaceOut(BaseModel):
    osm_id:          str
    name:            str
    category:        str
    latitude:        float
    longitude:       float
    distance_meters: float


class CompetitorSummaryOut(BaseModel):
    direct_count:      int
    related_count:     int
    total_businesses:  int
    competition_level: str
    density_per_sqkm:  float


class MarketOpportunityOut(BaseModel):
    competition_score:    float
    infrastructure_score: float
    accessibility_score:  float
    diversity_score:      float
    market_size_score:    float
    total:                float


class LocationSuitabilityOut(BaseModel):
    competition_score:    float
    infrastructure_score: float
    customer_proxy_score: float
    business_density:     float
    total:                float


class MarketInsightOut(BaseModel):
    icon:    str
    message: str
    level:   str


class MarketAnalysisOut(BaseModel):
    latitude:             float
    longitude:            float
    radius_km:            float
    location_name:        Optional[str]
    business_name:        str
    business_id:          str
    competitor_summary:   CompetitorSummaryOut
    opportunity:          MarketOpportunityOut
    suitability:          LocationSuitabilityOut
    nearby_places:        List[NearbyPlaceOut]
    direct_competitors:   List[NearbyPlaceOut]
    insights:             List[MarketInsightOut]
    recommendations:      List[str]
    disclaimer:           str


class NearbyResultOut(BaseModel):
    latitude:       float
    longitude:      float
    radius_km:      float
    total_places:   int
    places:         List[NearbyPlaceOut]
    categories:     Dict[str, int]
    from_cache:     bool
    error:          Optional[str]
    disclaimer:     str = (
        "Market intelligence results are based on available OpenStreetMap data "
        "and may be incomplete or outdated."
    )
